### PR TITLE
fix(challenge): correct terminology and inform of correct array copying

### PR DESCRIPTION
**Pull request by lasjorg**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

1. Change "Declare function arguments", to "Declare function parameters".

2. Make it clear that using variable-to-variable assignment for object/array copying does not work.

### CODE REVIEW

**Comment by lasjorg** [comment]
Do we actually teach why object/array copying works as it does in JS anywhere in the curriculum? We teach _how_ to copy an array, but do we teach why it is done like it is. Maybe I'm just not seeing where we teach this?

I didn't really feel like this was the proper place to get into any details about the subject so I just left it as an informational note.

**Comment by ieahleen** [comment]
Why don't you first open an issue so we can discuss about the better way of putting it?

**Comment by Sky020** [comment]
It is only mentioned, but not fully taught (from what I can see): https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-data-structures/copy-an-array-with-the-spread-operator

When I was going through the curriculum, I have a fuzzy memory that there was an external link to something like an fCC article going in-depth as to why.

I would agree with adding useful information around this, but not necessarily in this lesson, and not quite like this.

**Comment by RandellDawson** [comment]
I do not believe this is the right challenge to add this information. Maybe some similar could go in the challenge related to copying an array.

**Comment by nhcarrigan** [comment]
Should I create an issue to hold this discussion?

**Comment by RandellDawson** [comment]
We can continue to discuss here.

**Comment by nhcarrigan** [comment]
So [this challenge](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/functional-programming/avoid-mutations-and-side-effects-using-functional-programming) discusses the side effects of mutating an array with `splice`, but the challenge itself isn't really relevant. The [previous challenge](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/functional-programming/understand-the-hazards-of-using-imperative-code) is a bit more relevant, but that one is already a rather long wall of text. 

I do like @Sky020's proposal for adding it to the spread operator lesson, though.

**Comment by lasjorg** [comment]
Well, the first part needs correction no matter what, we do not declare an argument.

Secondly, I don't see why we can't have a note reminding the camper of how object copying works seeing as they need to create a copy on this challenge. I can make it shorter though.

Maybe just:

> **Note**: You can not create new copies (or clones) of objects and arrays by assigning the variable that holds them to a new variable.

**Comment by nhcarrigan** [comment]
I like the addition of the note for sure. 

It might be worth exploring the idea of adding more depth about this to an earlier challenge, too - but the note is perfect here IMO.

**Comment by lasjorg** [comment]
@nhcarrigan Oh I agree, this subject has to be covered somewhere, in more depth. At first, I was kind of just assuming I wasn't finding where in the curriculum it was covered.

I think the challenge on copying an array should cover this for one. Without an explanation for why copying is done like it is, it doesn't even make much sense.

I would also suggest that teaching functional programming and not teaching this concept to be a flaw. Avoiding object mutation is a big part of the "not mutating" concept.

**Comment by ojeytonwilliams** [comment]
> Note: remember that assigning an array to a new variable does not create a new array.  If you create an array assigned to `x` then assign it to `y`, you still have a single array.  For example `var x = [1,2]; var y = x; y[0] = 3; console.log(x); // [3,2]`

How about something like this for the note?

We could introduce the concept when we introduce [slice](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-data-structures/copy-array-items-using-slice), since that challenge isn't too wordy.

**Comment by ieahleen** [comment]
the `slice` challenge seems a good place

**Comment by lasjorg** [comment]
@ojeytonwilliams Not sure if it is getting too lengthy? But I'm all for giving a code example, I'm not sure what others have to say about this? If we give a lengthier note should it still be inline with point 1) or at the bottom on this own?

I also explicitly avoided using the word "remember" as I didn't find this information anywhere in the curriculum. So we have to make sure this information is given first.

> We could introduce the concept when we introduce [slice](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-data-structures/copy-array-items-using-slice), since that challenge isn't too wordy.

Yes, that seems like a good place.

Personally, I would like to have seen a new challenge added just before the array copying that explains this concept in some depth. So basically teaching a concept before teaching the method used. I know we can't do that (we can't add new challenges now, right?) so I guess the slice challenge is the best place for this.

**Comment by lasjorg** [comment]
Side note: We should probably avoid using pass-by-reference when teaching this. I believe this is a bit of a misnomer and that JS always does pass-by-value (the value can be an address) and what is sometimes referred to as [Call by sharing](https://en.wikipedia.org/wiki/Evaluation_strategy#Call_by_sharing). The [article](https://codeburst.io/javascript-passing-by-value-vs-reference-explained-in-plain-english-8d00fd06a47c) linked to from the Hints page does not really make that distinction.

http://dmitrysoshnikov.com/ecmascript/chapter-8-evaluation-strategy/

**Comment by ojeytonwilliams** [comment]
@lasjorg yep, my version might be a bit too long. Feel free to use whatever parts of it you like and drop the rest. Thinking about it,  "remember" could just be dropped:

> Note: assigning an array...

seems fine.

**Comment by RandellDawson** [comment]
@lasjorg Any more progress on this one?

**Comment by lasjorg** [comment]
@RandellDawson I don't know. Nobody really gave any feedback as to how we should best handle the part about object/array copying.

- Do we give an example, or just make it a simple reminder note?
- If we give a code example, what should it be and where should it be placed?

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/refactor-global-variables-out-of-functions.md [hunk 1]
```diff
@@ -8,8 +8,8 @@ forumTopicId: 301235
 ## Description
 <section id='description'>
 So far, we have seen two distinct principles for functional programming:
-1) Don't alter a variable or object - create new variables and objects and return them if need be from a function.
-2) Declare function arguments - any computation inside a function depends only on the arguments, and not on any global object or variable.
+1) Don't alter a variable or object - create new variables and objects and return them if need be from a function. Note, you can not create new copies (or clones) of objects and arrays by assigning the variable that holds them to a new variable. So for example <code>const arrayCopy = originalArray;</code> does <strong>not</strong> create a new copy. Instead, use one of the array copy techniques shown previously in the curriculum.
```
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/refactor-global-variables-out-of-functions.md:11]:
```suggestion
1) Don't alter a variable or object - create new variables and objects and return them if need be from a function. Hint: using something like `var newArr = arrVar`, where `arrVar` is an array will simply create a reference to the existing variable and not a copy. So changing a value in `newArr` would change the value in `arrVar`.
```

**Review by moT01 — COMMENTED** [review]
1) Don't alter a variable or object - create new variables and objects and return them if need be from a function. Hint: using something like `var newArr = arrVar`, where `arrVar` is an array will simply create a reference to the existing variable and not a copy. So changing a value in `newArr` would change the value in `arrVar`.

I dunno - I tried to come up with an alternative phrasing, it's okay. Really, I suppose I would be fine adding this how it is or using Oliver's suggestion. Or anything alone these lines really. I vote to let @lasjorg decide and then we just go with it.

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Review by RandellDawson — APPROVED** [review]
LGTM

**Comment by RandellDawson** [comment]
Thank you for your contribution to the page! 👍
We're happy to accept these changes, and look forward to future contributions. 🎉
